# Singularity example

*Last edited: 2024-06-11*

Singularity (now Apptainer) example. This Notebook runs on SDumont.

References:

- <https://github.com/LucasFernando-aes/SDumont-UserPOV/blob/main/singularity.md>
- <https://github.com/apptainer/singularity/releases/tag/v3.8.5>
- <https://apptainer.org/>
- <https://guiesbibtic.upf.edu/recerca/hpc/singularity>
- <https://pawseysc.github.io/hpc-container-training/13-singularity-intro/>
- <https://epcced.github.io/2020-12-08-Containers-Online/aio/index.html>
- <https://singularity-tutorial.github.io/02-basic-usage/>
- <https://hsf-training.github.io/hsf-training-singularity-webpage/>
- <https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch/tags>
- <https://docs.sylabs.io/guides/3.8/user-guide/>

In November 2021 the Singularity open source project joined the Linux Foundation [and was renamed](https://en.wikipedia.org/wiki/Singularity_(software)) to Apptainer.

Check for related modules:

In [3]:
! module avail -t 2>&1 | grep -i docker

In [2]:
! module avail -t 2>&1 | grep -i apptainer

In [1]:
! module avail -t 2>&1 | grep -i singularity

parabricks/2.5.0_singularity_sequana
parabricks/3.0_singularity_sequana


Check the available Singularity version:

In [4]:
! singularity --version

singularity version 3.8.5-1.el7


Download example container:

In [8]:
! singularity pull --force library://godlovedc/funny/lolcow

INFO:    Using cached image


In [9]:
! ls lolcow_latest.sif

lolcow_latest.sif


The following cell ran in a terminal, starting a container and running an interactive shell inside it:

Executing a command within the container, via the host's command line:

In [1]:
! singularity exec lolcow_latest.sif cat /etc/os-release

NAME="Ubuntu"
VERSION="16.04.5 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.5 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial


In [2]:
! singularity exec lolcow_latest.sif cowsay moo

 _____
< moo >
 -----
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||


Another example, downloading the PyTorch container:

<https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch/tags>

nvcr.io/nvidia/pytorch:22.03-py3

In [1]:
%%bash
export SINGULARITY_DOCKER_USERNAME='$oauthtoken'
export SINGULARITY_DOCKER_PASSWORD=<key>
time singularity pull pytorch22.03.sif docker://nvcr.io/nvidia/pytorch:22.03-py3

INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob sha256:4d32b49e2995210e8937f0898327f196d3fcc52486f0be920e8b2d65f150a7ab
Copying blob sha256:45893188359aca643d5918c9932da995364dc62013dfa40c075298b1baabece3
Copying blob sha256:5ad1f2004580e415b998124ea394e9d4072a35d70968118c779f307204d6bd17
Copying blob sha256:6ddc1d0f91833b36aac1c6f0c8cea005c87d94bab132d46cc06d9b060a81cca3
Copying blob sha256:4cc43a803109d6e9d1fd35495cef9b1257035f5341a2db54f7a1940815b6cc65
Copying blob sha256:e94a4481e9334ff402bf90628594f64a426672debbdfb55f1290802e52013907
Copying blob sha256:3e7e4c9bc2b136814c20c04feb4eea2b2ecf972e20182d88759931130cfb4181
Copying blob sha256:9463aa3f56275af97693df69478a2dc1d171f4e763ca6f7b6f370a35e605c154
Copying blob sha256:a4a0c690bc7da07e592514dccaa26098a387e8457f69095e922b6d73f7852502
Copying blob sha256:59d451175f6950740e26d38c322da0ef67cb59da63181eb32996f752ba8a2f17
Copying blob sha256:eaf45e9f32d1f5a9983945a1a9f

CalledProcessError: Command 'b"export SINGULARITY_DOCKER_USERNAME='$oauthtoken'\nexport SINGULARITY_DOCKER_PASSWORD=MzZ0ZmRtcWFkMXZsaWNnN2NmYTVsZTk4Nm06OTcwZGNhYjYtNjcyNy00OTQ0LTgxYzMtZDBmZmEwNDZmZWU1\ntime singularity pull pytorch22.03.sif docker://nvcr.io/nvidia/pytorch:22.03-py3\n"' returned non-zero exit status 255.

It is not possible to download for this container, a timeout occurs on the login node, and on an interactive node an error message appears when authenticating in the Nvidia repo. The interactive execution node is configured using:

    salloc --partition=sequana_gpu_shared --nodes=1 --ntasks=2 --time=30 --job-name test01
    ssh sdumontxxxx
    ...
    exit
    exit

The solution is to download to the local machine using Docker, save as .tar, and transfer to SDumont using scp or syncthing. On the local machine:

    docker login nvcr.io    # USERNAME&PASSWORD
    docker pull nvcr.io/nvidia/pytorch:22.03-py3
    docker save nvcr.io/nvidia/pytorch:22.03-py3  > local.tar
    transfer (to the image/ directory)

On the login node it is not possible to convert the image to SIF format:

In [1]:
%%bash
cd ../image
ls
time singularity build local_tar.sif docker-archive://local.tar

local.tar


INFO:    Starting build...
Getting image source signatures
Copying blob sha256:867d0767a47c392f80acb51572851923d6d3e55289828b0cd84a96ba342660c7
Copying blob sha256:01e996931197162c4d65bfc6867f243df01e5f70f877c1c9beccd6978297e643
Copying blob sha256:2ff0ade8d3c97b175cf6be4658d30fc655fa86b0198224c8f5b18e39cdd97e5f
Copying blob sha256:fec6965e7a6b67b61a1337a169d91dcbb93ebeee67bb5d38a241f7fa4fa048ea
Copying blob sha256:83cdade3c9b5e8313ab6d74bcfcc0492a4ae4b9fbf3d09a35ab52abaaf492d9d
Copying blob sha256:a060c5cefec7f091c555e77e46d0130a49dfc39023365e216c154613603f456a
Copying blob sha256:2df8c0a32afe2d088aafe59dc368643b14ddd15f9673cac769548c28ec384e04
Copying blob sha256:899455397741dac157e85ecf485ec909e52ce1ada5d37853c93544c13e7f5f81
Copying blob sha256:2f175b794573ab5e137bd2f5f52a2c1716bb5287ca8755ea8ad166fa9b1ee898
Copying blob sha256:85f49f4e6923e73ae6c3ab79d4b7f1f8156141952f3f69535ca49f6a543f9e24
Copying blob sha256:6a1014d46250ad84e998df518cecf334cf4dc3255d8da2107f5b14adb0137638
Copyin

CalledProcessError: Command 'b'cd ../image\nls\nsingularity build local_tar.sif docker-archive://local.tar\n'' returned non-zero exit status 255.

You need to copy the .tar.gz to scratch, start an interactive shell, and then convert to SIF:

In [2]:
! cp local.tar /scratch/<user-dir>/container/

In a shell:

    salloc --partition=sequana_gpu_shared --nodes=1 --ntasks=2 --time=30 --job-name test01
    ssh sdumontxxxx
    cd ...
    singularity build pytorch22.03.sif docker-archive://local.tar
    ...

Running on the sdumont18:

In [11]:
! singularity exec --nv /scratch/ampemi/<urser>/container/pytorch22.03.sif nvidia-smi

Tue Jun 11 14:23:27 2024       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  Off  | 00000000:3B:00.0 Off |                    0 |
| N/A   41C    P0    38W / 250W |      0MiB / 32510MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-PCIE...  Off  | 00000000:5E:00.0 Off |                    0 |
| N/A   

Running interactively in a shell:

    $ singularity shell --nv pytorch22.03.sif
    Singularity> nvidia-smi
    Tue Jun 11 11:15:49 2024       
    +-----------------------------------------------------------------------------+
    | NVIDIA-SMI 470.82.01    Driver Version: 470.82.01    CUDA Version: 11.6     |
    |-------------------------------+----------------------+----------------------+
    | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    |                               |                      |               MIG M. |
    |===============================+======================+======================|
    |   0  Tesla V100-PCIE...  Off  | 00000000:3B:00.0 Off |                    0 |
    | N/A   45C    P0    39W / 250W |      0MiB / 32510MiB |      0%      Default |
    |                               |                      |                  N/A |
    +-------------------------------+----------------------+----------------------+
    |   1  Tesla V100-PCIE...  Off  | 00000000:5E:00.0 Off |                    0 |
    | N/A   36C    P0    37W / 250W |      0MiB / 32510MiB |      0%      Default |
    |                               |                      |                  N/A |
    +-------------------------------+----------------------+----------------------+
    |   2  Tesla V100-PCIE...  Off  | 00000000:86:00.0 Off |                    0 |
    | N/A   35C    P0    35W / 250W |      0MiB / 32510MiB |      0%      Default |
    |                               |                      |                  N/A |
    +-------------------------------+----------------------+----------------------+
    |   3  Tesla V100-PCIE...  Off  | 00000000:AF:00.0 Off |                    0 |
    | N/A   36C    P0    35W / 250W |      0MiB / 32510MiB |      2%      Default |
    |                               |                      |                  N/A |
    +-------------------------------+----------------------+----------------------+
                                                                                   
    +-----------------------------------------------------------------------------+
    | Processes:                                                                  |
    |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
    |        ID   ID                                                   Usage      |
    |=============================================================================|
    |  No running processes found                                                 |
    +-----------------------------------------------------------------------------+
    Singularity> 

Check PyTorch:

```python
$ singularity shell --nv pytorch22.03.sif
Singularity> python
Python 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:42:07) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> x = torch.rand(5, 3)
>>> print(x)
tensor([[0.7518, 0.3518, 0.3545],
        [0.5821, 0.3066, 0.6131],
        [0.0897, 0.5199, 0.5225],
        [0.4321, 0.3631, 0.0693],
        [0.1293, 0.1515, 0.0600]])
>>>
>>> import torch
>>> torch.cuda.is_available()
True
```